proposals/list-patterns.md

```diff
@@ -121,6 +172,141 @@ expr.Length is >= 2
+
+Although a helper type is not necessary, it helps simplify and illustrate the logic.
+
+```
+class ListPatternHelper
+{
+  // Notes: 
+  // We could inline this logic to avoid creating a new type and to handle the pattern-based enumeration scenarios.
+  // We may only need one element in start buffer, or maybe none at all, if we can control the order of checks in the patterns DAG.
+  // We could emit a count check for a non-terminal `..` and economize on count checks a bit.
+  private EnumeratorType enumerator;
+  private int count;
+  private ElementType[] startBuffer;
+  private ElementType[] endCircularBuffer;
+
+  public ListPatternHelper(EnumerableType enumerable, int startPatternsCount, int endPatternsCount)
+  {
+    count = 0;
+    enumerator = enumerable.GetEnumerator();
+    startBuffer = startPatternsCount == 0 ? null : new ElementType[startPatternsCount];
+    endCircularBuffer = endPatternsCount == 0 ? null : new ElementType[endPatternsCount];
+  }
+
+  // targetIndex = -1 means we want to enumerate completely
+  private int MoveNextIfNeeded(int targetIndex)
+  {
+    int startSize = startBuffer?.Length ?? 0;
+    int endSize = endCircularBuffer?.Length ?? 0;
+    Debug.Assert(targetIndex == -1 || (targetIndex >= 0 && targetIndex < startSize));
+
+    while ((targetIndex == -1 || count <= targetIndex) && enumerator.MoveNext())
+    {
+      if (count < startSize)
+        startBuffer[count] = enumerator.Current;
+
+      if (endSize > 0)
+        endCircularBuffer[count % endSize] = enumerator.Current;
+
+      count++;
+    }
+
+    return count;
+  }
+
+  public int Count()
+  {
+    return MoveNextIfNeeded(-1);
+  }
+
+  // fulfills the role of `[index]` for start elements when enough elements are available
+  public bool TryGetStartElement(int index, out ElementType value)
+  {
+    Debug.Assert(startBuffer is not null && index >= 0 && index < startBuffer.Length);
+    MoveNextIfNeeded(index);
+    if (count > index)
+    {
+      value = startBuffer[index];
+      return true;
+    }
+    value = default;
+    return false;
+  }
+
+  // fulfills the role of `[^hatIndex]` for end elements when enough elements are available
+  public ElementType GetEndElement(int hatIndex)
+  {
+    Debug.Assert(endCircularBuffer is not null && hatIndex > 0 && hatIndex <= endCircularBuffer.Length);
+    int endSize = endCircularBuffer.Length;
+    Debug.Assert(endSize > 0);
+    return endCircularBuffer[(count - hatIndex) % endSize];
+  }
+}
+```
+
+`collection is [3]` is lowered to
+```
+@{
+  var helper = new ListPatternHelper(collection, 0, 0);
+
+  helper.Count() == 3
+}
+```
+
+`collection is { 0, 1 }` is lowered to
+```
+@{
+  var helper = new ListPatternHelper(collection, 2, 0);
+
+  helper.TryGetStartElement(index: 0, out var element0) && element0 is 0 &&
+  helper.TryGetStartElement(1, out var element1) && element1 is 1 &&
+  helper.Count() == 2
```

Comment:
Same comment about enumerable length.

Reply:
Thanks. We only need to check we're at the end of the enumeration.